lib/transport/control.ts

```diff
@@ -66,17 +66,17 @@ export interface Subscribe {
 	namespace: string
 	name: string
 
-	start_group: Location
-	start_object: Location
-	end_group: Location
-	end_object: Location
+	location: Location
 
 	params?: Parameters
 }
 
 export interface Location {
-	mode: "none" | "absolute" | "latest" | "future"
-	value?: number // ignored for type=none, otherwise defaults to 0
+	mode: "LatestGroup" | "LatestObject" | "AbsoluteStart" | "AbsoluteRange"
+	start_group: number // ignored except for AbsoluteStart, AbsoluteRange
+	start_object: number // ignored except for AbsoluteStart, AbsoluteRange
+	end_group: number // ignored except for AbsoluteRange
+	end_object: number // ignored except for AbsoluteRange
```

Comment:
This is the part where it most feels like I don't know my way around TypeScript as well as I'd like. I could make these optional as before, but having that decoupled (at the type level) from what we know about the modes looked like it might make for some very verbose `undefined` checking.

In the interest of time, I went with something that seemed _adequate_ for now.

Reply:
```typescript
export type Location = LatestGroup | LatestObject | AbsoluteStart | AbsoluteEnd;

export interface LatestGroup {
  mode: "latest_group",
}

export interface LatestObject {
  mode: "latest_object",
}

export interface AbsoluteStart {
  mode: "absolute_start",
  start_group: number,
  start_object: number,
}

export interface AbsoluteRange {
  mode: "absolute_range",
  start_group: number,
  start_object: number,
  end_group: number,
  end_object: number,
}
```

Reply:
Ah! Yes, I was looking for something like Rust's enum's with struct-like members and this looks like it. Thanks!